privacy-protections/storage-partitioning/main.js

```diff
@@ -0,0 +1,253 @@
+
+        span.innerHTML = `${type}: ${JSON.stringify(allRetrievals.get(api)[type], null, 2)}`;
+        li.appendChild(span);
+        return li;
+    }
+
+    function getIcon (result) {
+        if (result === 'pass' || result === 'unsupported') {
+            return '✅';
+        } else if (result === 'fail') {
+            return '❌';
+        } else if (result === 'error') {
+            return '⚠️';
+        } else {
+            return '?';
+        }
+    }
+
+    for (const api of allRetrievals.keys()) {
+        const result = testResults.get(api);
+
+        const li = document.createElement('li');
+        li.id = `test-${api.replace(' ', '-')}`;
+        li.innerHTML = `${getIcon(result)} ${api} - ${result}<ul></ul>`;
+
+        const ul = li.getElementsByTagName('ul')[0];
+        ul.appendChild(getLiFromResults(api, 'same-site'));
+        ul.appendChild(getLiFromResults(api, 'cross-site'));
+
+        testsDetailsElement.appendChild(li);
+    };
+
+    testsSummaryDiv.innerText = `Retrieved data from ${allRetrievals.size} storage mechanisms.`;
+
+    function addTestResult (testId, value) {
+        results.results.push({
+            id: testId,
+            value: value
+        });
+    }
+
+    testResults.forEach((status, api) => {
+        addTestResult(api, status);
+    });
+
+    downloadButton.removeAttribute('disabled');
+}
+
+// From: https://stackoverflow.com/a/51321724
+class DefaultMap extends Map {
+    get (key) {
+        if (!this.has(key)) this.set(key, this.default());
+        return super.get(key);
+    }
+
+    constructor (defaultFunction, entries) {
+        super(entries);
+        this.default = defaultFunction;
+    }
+}
+
+async function runTests () {
+    const sessionId = uuidv4();
+
+    // Update favicon (must happen in top-level frame)
+    const faviconElement = document.getElementById('favicon');
+    const faviconURL = new URL('/partitioning/resource', window.location.origin);
+    faviconURL.searchParams.set('fileType', 'favicon');
+    faviconURL.searchParams.set('key', sessionId);
+    faviconElement.setAttribute('href', faviconURL.href);
+
+    // Open the test tab where all tests will be run
+    const testURL = new URL('/privacy-protections/storage-partitioning/testWindow.html', window.location.origin);
+    testURL.searchParams.set('sessionId', sessionId);
+    testURL.searchParams.set('isLocalTest', isLocalTest);
+
+    // Ideally we'd use noopener here as some browsers (e.g., Firefox) keep tabs that
+    // have references to each other in the same "agent cluster", which impacts the
+    // scope of storage. This is why you'll see consistent sessionStorage across tabs
+    // that have references to each other in Firefox. Unfortuantely the test tab can't
+    // be closed programatically unless it has an opener reference.
+    // See: https://textslashplain.com/2021/02/04/window-close-restrictions/
+    window.open(testURL, '_blank');
```

Comment:
I remember some browsers considering to assume 'noopener' for `target=_blank`. You haven't observed it though, right?

Reply:
I'll verify. Arthur and I found that noopener explains some of the differences we see for Chrome re: https://app.asana.com/0/0/1201683514701538/f, so it might make sense to have two versions of the tests--once which preserves the opener and one which severs it.

Reply:
I haven't observed this, and decided not to add the noopener version of the test for now. I'll make a not in Asana to consider it.